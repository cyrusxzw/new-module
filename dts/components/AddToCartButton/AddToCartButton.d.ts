import { FC } from 'react';
import type { AddToCartButtonProps } from './AddToCartButton.types';
declare const AddToCartButton: FC<AddToCartButtonProps>;
export { AddToCartButton };
