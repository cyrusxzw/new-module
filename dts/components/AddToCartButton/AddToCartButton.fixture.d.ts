declare const AddToCartButtonFixture: {
  className: string;
  copy: {
    cartAction: string;
    updateNotification: string;
    outOfStock: {
      label: string;
      title: string;
    };
  };
  dataTestRef: string;
  isEnabled: boolean;
  isFullWidth: boolean;
  theme: string;
  product: {
    size: string;
    sku: string;
    description: string;
    cartDisclaimer: string;
    definitionList: {
      term: string;
      id: string;
      description: string;
    }[];
    id: string;
    ingredients: string;
    keyIngredient: string;
    productName: string;
    upSellProduct: {
      image: {
        altText: string;
        small: string;
      };
      name: string;
      url: string;
    };
  };
  variants: {
    cartDisclaimer: string;
    size: string;
    sku: string;
    price: string;
    isInStock: boolean;
    image: {
      altText: string;
      sizes: {
        large: string;
        medium: string;
        small: string;
      };
    };
  }[];
  outOfStockVariant: {
    cartDisclaimer: string;
    size: string;
    sku: string;
    price: string;
    isInStock: boolean;
    image: {
      altText: string;
      sizes: {
        large: string;
        medium: string;
        small: string;
      };
    };
  }[];
  alternateActionVariant: {
    alternateAction: {
      url: string;
      label: string;
    };
    cartDisclaimer: string;
    size: string;
    sku: string;
    price: string;
    isInStock: boolean;
    image: {
      altText: string;
      sizes: {
        large: string;
        medium: string;
        small: string;
      };
    };
  }[];
};
export { AddToCartButtonFixture };
