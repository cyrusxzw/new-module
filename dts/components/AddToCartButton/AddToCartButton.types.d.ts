import type { Themes } from '~/types';
declare type AddToCartButtonProps = {
  className?: string;
  copy?: {
    cartAction?: string;
    updateNotification?: string;
    outOfStock?: {
      label?: string;
      title?: string;
    };
  };
  dataTestRef: string;
  isEnabled?: boolean;
  isFullWidth?: boolean;
  theme?: Themes;
};
export type { AddToCartButtonProps };
