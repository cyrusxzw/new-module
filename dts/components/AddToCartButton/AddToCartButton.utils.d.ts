declare const useDelayedDisabled: (
  shouldShowUpdateSuccessMessage: boolean,
) => boolean;
export { useDelayedDisabled };
