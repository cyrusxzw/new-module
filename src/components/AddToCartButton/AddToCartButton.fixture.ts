const AddToCartButtonFixture = {
  className: '',
  copy: {
    cartAction: 'Add to your cart',
    updateNotification: 'Added to your cart',
    outOfStock: {
      label: 'Out of Stock',
      title: 'Out of stock',
    },
  },
  dataTestRef: 'DATA_TEST_REF',
  isEnabled: true,
  isFullWidth: true,
  theme: 'dark',
  product: {
    size: '50 mL',
    sku: 'ARD33',
    description:
      'Nunc sagittis ultrices dignissim. Nullam dignissim nisi nec dolor blandit convallis. Suspendisse quis facilisis arcu, nec feugiat ligula.',
    cartDisclaimer:
      'Quisque in felis id tellus condimentum scelerisque. Aliquam quis vestibulum felis, quis porta tellus',
    definitionList: [
      {
        term: 'Lorem ipsum dolor',
        id: 'Lorem-ipsum-dolor',
        description: 'Lorem ipsum dolort, consectetur adipiscing elit.',
      },
      {
        term: 'Ut consectetur mi',
        id: 'Ut-consectetur-mi',
        description: 'Ut consectetur, vitae libero imperdiet id.',
      },
    ],
    id: 'Lorem-ipsum-dolor',
    ingredients:
      'Maecenas mattis, urna vel commodo vehicula, diam tellus vestibulum urna, in lobortis lacus quam et arcu. Nunc arcu neque, ullamcorper et eros vitae, scelerisque posuere quam. Nunc sagittis ultrices dignissim. Nullam dignissim nisi nec dolor blandit convallis. Suspendisse quis facilisis arcu, nec feugiat ligula. Vestibulum quis malesuada nisi, a finibus eros. Praesent et vestibulum diam. Nam tristique viverra odio, ut tincidunt diam placerat eu. Proin eu magna imperdiet vulputate ex sit amet, laoreet neque. Praesent rhoncus ex quis elit vehicula, et tincidunt nisi scelerisque. Nullam ultrices scelerisque ex at efficitur.',
    keyIngredient: 'Maecenas mattis, urna vel commodo vehicula',
    productName: 'Lorem ipsum dolor',
    upSellProduct: {
      image: {
        altText: 'Product Extract',
        small:
          'https://www.aesop.com/medias/Aesop-Kits-Gift-Kits-2019-20-The-Reveller-with-Product-Small-781x524px.png?context=bWFzdGVyfGltYWdlc3wxMTMwMjN8aW1hZ2UvcG5nfGltYWdlcy9oZjgvaDI1LzkxNTAzNjg5NDAwNjIucG5nfDllZjNkMjBkNjZhZTI0OGRhZWZhOGJlMzA5NDM3OTg4ZGFlNjE1MWY1ZDg2ZmU4MjUwZmEwOWMwMmRkNzhmNDg',
      },
      name: 'The Reveller Fragrant hand and body care',
      url: '/',
    },
  },
  variants: [
    {
      cartDisclaimer: 'Praesent rhoncus ex quis elit vehicula',
      size: '50 mL - In stock',
      sku: 'ARD33',
      price: '$26.45',
      isInStock: true,
      image: {
        altText: 'alt-tag-ARD33',
        sizes: {
          large:
            'https://www.aesop.com/medias/Aesop-Kits-Gift-Kits-2020-21-The-Ardent-Nomad-Hybris-Large-1584x962px.png?context=bWFzdGVyfGltYWdlc3w0MTI1MzB8aW1hZ2UvcG5nfGltYWdlcy9oZjcvaDVlLzk1NjMwMjE3MDUyNDYucG5nfGUyYjQyMjlmMDQ0ZjdlOTRjMzAyNDczYzA2NGQ1Yzk0YWI4ZWQ0OTBkMmQzYWNjNWMxMjZjMDU1YTZjZDFhMzE',
          medium:
            'https://www.aesop.com/medias/Aesop-Kits-Gift-Kits-2020-21-The-Ardent-Nomad-Hybris-Large-1584x962px.png?context=bWFzdGVyfGltYWdlc3w0MTI1MzB8aW1hZ2UvcG5nfGltYWdlcy9oZjcvaDVlLzk1NjMwMjE3MDUyNDYucG5nfGUyYjQyMjlmMDQ0ZjdlOTRjMzAyNDczYzA2NGQ1Yzk0YWI4ZWQ0OTBkMmQzYWNjNWMxMjZjMDU1YTZjZDFhMzE',
          small:
            'https://www.aesop.com/medias/Aesop-Kits-Gift-Kits-2020-21-The-Ardent-Nomad-Hybris-Large-1584x962px.png?context=bWFzdGVyfGltYWdlc3w0MTI1MzB8aW1hZ2UvcG5nfGltYWdlcy9oZjcvaDVlLzk1NjMwMjE3MDUyNDYucG5nfGUyYjQyMjlmMDQ0ZjdlOTRjMzAyNDczYzA2NGQ1Yzk0YWI4ZWQ0OTBkMmQzYWNjNWMxMjZjMDU1YTZjZDFhMzE',
        },
      },
    },
    {
      cartDisclaimer: 'Pet tincidunt nisi scelerisque',
      size: '100 mL - Not in stock',
      sku: 'ARD32',
      price: '$86.97',
      isInStock: false,
      image: {
        altText: 'alt-tag-ARD32',
        sizes: {
          large: './assets/images/Product/variant-two-large.png',
          medium: './assets/images/Product/variant-two-medium.png',
          small: './assets/images/Product/variant-two-small.png',
        },
      },
    },
  ],
  outOfStockVariant: [
    {
      cartDisclaimer: 'Nunc sagittis ultrices dignissim',
      size: '50 ml',
      sku: 'ARD33',
      price: '$26.45',
      isInStock: false,
      image: {
        altText: 'alt-tag-ARD33',
        sizes: {
          large: './assets/images/Product/variant-one-large.png',
          medium: './assets/images/Product/variant-one-medium.png',
          small: './assets/images/Product/variant-one-small.png',
        },
      },
    },
  ],
  alternateActionVariant: [
    {
      alternateAction: {
        url: '/',
        label: 'alternateActionVariant alternateAction label',
      },
      cartDisclaimer: 'alternateActionVariant cartDisclaimer',
      size: 'size ml',
      sku: 'alternateActionVariant-sku',
      price: '$12.34',
      isInStock: false,
      image: {
        altText: 'alternateActionVariant image altText',
        sizes: {
          large: './assets/images/Product/variant-one-large.png',
          medium: './assets/images/Product/variant-one-medium.png',
          small: './assets/images/Product/variant-one-small.png',
        },
      },
    },
  ],
};

export { AddToCartButtonFixture };
