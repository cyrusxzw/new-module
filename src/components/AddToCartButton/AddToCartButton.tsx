import React from 'react';
import cx from 'classnames';
import { useAddToCartContext, useVariantSelectContext } from '~/contexts';
import { HYPERLINK_STYLE_TYPES } from '~/constants';
import { Button } from '~/components/Button';
import { Loading } from '~/components/Loading';
import { Hyperlink } from '~/components/Hyperlink';
import { useDelayedDisabled } from './AddToCartButton.utils';
import type { AddToCartButtonType } from './AddToCartButton.types';
import styles from './AddToCartButton.module.css';

const AddToCartButton: AddToCartButtonType = ({
  className,
  copy,
  dataTestRef = 'ADD_TO_CART',
  isEnabled = true,
  isFullWidth = true,
  theme = 'dark',
}) => {
  const addToCartContext = useAddToCartContext();
  const { selectedVariant } = useVariantSelectContext();
  const {
    errorMessage,
    hasError,
    isLoading,
    isUpdateSuccessful,
  } = addToCartContext;
  const shouldShowUpdateSuccessMessage = !isLoading && isUpdateSuccessful;
  const isDelayedDisabled = useDelayedDisabled(shouldShowUpdateSuccessMessage);
  const updateNotificationLabel = copy?.updateNotification;

  if (!selectedVariant) return null;

  const {
    alternateAction,
    isInStock,
    isSellable,
    price,
    sku,
  } = selectedVariant;

  const cartActionText = copy?.cartAction
    ? `${copy?.cartAction} â€” ${price}`
    : price;
  const postTaxPriceText = copy?.postTaxPrice ? ` ${copy?.postTaxPrice}` : '';
  const cartActionLabel = cartActionText + postTaxPriceText;

  const classSet = cx(
    styles.base,
    { [styles.fullWidth]: isFullWidth },
    className,
  );

  const handleOnClick = () => {
    const { dispatch, onClick } = addToCartContext;

    onClick(sku, dispatch);
  };

  if (!!alternateAction?.url && !!alternateAction?.label) {
    return (
      <Hyperlink
        className={classSet}
        dataTestRef={dataTestRef}
        hasTargetInNewWindow={alternateAction?.openInANewWindow}
        isAlternate={true}
        style={HYPERLINK_STYLE_TYPES.INTERNAL_NO_ICON_BUTTON_LINK}
        theme={theme}
        title={alternateAction?.label}
        url={alternateAction?.url}
      >
        {alternateAction?.label}
      </Hyperlink>
    );
  }

  if (hasError) {
    /** @TODO Handle errors thrown by handleOnClick */
    console.error('Add To Cart button updateError: ', errorMessage); // eslint-disable-line
  }

  const isButtonEnabled =
    !isLoading &&
    !!price &&
    !!sku &&
    isInStock &&
    !!sku &&
    isEnabled &&
    !hasError &&
    !isDelayedDisabled &&
    isSellable;

  const labelClassName = cx(
    styles.label,
    { [styles.hideLabel]: isLoading },
    { [styles.showSuccessMessage]: shouldShowUpdateSuccessMessage },
  );

  return (
    <Button
      className={classSet}
      dataTestRef={dataTestRef}
      isAlternate={true}
      isEnabled={isButtonEnabled}
      onClick={handleOnClick}
      theme={theme}
      title={isInStock ? cartActionLabel : copy?.outOfStock?.title}
    >
      {isLoading && (
        <Loading
          className={styles.loading}
          data-test-ref={`${dataTestRef}_UPDATING`}
          isLoading={true}
          theme={theme === 'dark' ? 'light' : 'dark'}
        />
      )}
      <span
        className={labelClassName}
        data-test-ref={
          shouldShowUpdateSuccessMessage ? `${dataTestRef}_SUCCESS` : undefined
        }
      >
        {shouldShowUpdateSuccessMessage ? (
          <span>{updateNotificationLabel}</span>
        ) : (
          <span> </span>
        )}
        <span>{isInStock ? cartActionLabel : copy?.outOfStock?.label}</span>
      </span>
    </Button>
  );
};

export { AddToCartButton };
