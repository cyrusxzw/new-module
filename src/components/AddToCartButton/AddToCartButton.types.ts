import type { Themes } from '~/types';

type AddToCartButtonProps = {
  className?: string;
  copy?: {
    cartAction?: string;
    postTaxPrice?: string;
    updateNotification?: string;
    outOfStock?: {
      label?: string;
      title?: string;
    };
  };
  dataTestRef: string;
  isEnabled?: boolean;
  isFullWidth?: boolean;
  theme?: Themes;
};

export type { AddToCartButtonProps };
