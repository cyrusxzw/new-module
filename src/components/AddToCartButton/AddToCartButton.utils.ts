import { useRef, useEffect, useState } from 'react';

const useDelayedDisabled = (
  shouldShowUpdateSuccessMessage: boolean,
): boolean => {
  const [isDelayedDisabled, setIsDelayedDisabled] = useState(false);
  const timeout = useRef(null);

  useEffect(() => {
    if (shouldShowUpdateSuccessMessage) {
      setIsDelayedDisabled(true);

      timeout.current = setTimeout(() => {
        setIsDelayedDisabled(false);
      }, 2500);
    }

    return function cleanup() {
      clearTimeout(timeout.current);
    };
  }, [shouldShowUpdateSuccessMessage]);

  return isDelayedDisabled;
};

export { useDelayedDisabled };
