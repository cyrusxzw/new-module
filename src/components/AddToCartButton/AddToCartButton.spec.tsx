import React from 'react';
import PropTypes from 'prop-types';
import { render, act, screen } from '@testing-library/react';
import userEvent from '@testing-library/user-event';
import {
  AddToCartContextProvider,
  ProductDetailContextProvider,
  VariantSelectContextProvider,
} from '~/contexts';
import { AddToCartButton } from './AddToCartButton';
import { AddToCartButtonFixture as fixture } from './AddToCartButton.fixture';
import { mockAddToCartButtonOnClick } from './__mocks__/AddToCartButton.onClick';

const AddToCartButtonWithProviders = ({ onClick, product, variants, copy }) => {
  return (
    <AddToCartContextProvider onClick={onClick}>
      <ProductDetailContextProvider product={product}>
        <VariantSelectContextProvider variants={variants}>
          <AddToCartButton copy={copy} dataTestRef="ADD_TO_CART" />
        </VariantSelectContextProvider>
      </ProductDetailContextProvider>
    </AddToCartContextProvider>
  );
};

AddToCartButtonWithProviders.propTypes = {
  onClick: PropTypes.func,
  product: PropTypes.object,
  variants: PropTypes.array,
  copy: PropTypes.object,
  copyJP: PropTypes.object,
};

describe('<AddToCartButton />', () => {
  const handleOnClick = jest.fn();
  const product = fixture.product;
  const variants = fixture.variants;
  const outOfStockVariant = fixture.outOfStockVariant;
  const alternateActionVariant = fixture.alternateActionVariant;
  const copy = fixture.copy;
  const copyJP = fixture.copyJP;

  beforeAll(() => {
    jest.spyOn(global.console, 'error').mockImplementation(() => {
      return;
    });
  });

  afterAll(() => {
    (global.console.error as jest.Mock).mockRestore();
  });

  afterEach(() => {
    jest.clearAllMocks();
    jest.useRealTimers();
  });

  it('should render base component correctly', () => {
    const { container } = render(
      <AddToCartButtonWithProviders
        copy={copy}
        onClick={mockAddToCartButtonOnClick}
        product={product}
        variants={variants}
      />,
    );

    const button = screen.getByRole('button', {
      name: `${fixture.copy.cartAction} — ${fixture.variants[0].price}`,
    });

    expect(button).toBeInTheDocument();
    expect(container).toMatchSnapshot();
  });

  it('should render add to cart button with text to indicate tax inclusion for jp region correctly', () => {
    const { container } = render(
      <AddToCartButtonWithProviders
        copy={copyJP}
        onClick={mockAddToCartButtonOnClick}
        product={product}
        variants={variants}
      />,
    );

    const button = screen.getByRole('button', {
      name: `${fixture.copyJP.cartAction} — ${fixture.copyJP.price}${' '}${
        fixture.copyJP.postTaxPrice
      }`,
    });
    expect(button).toBeInTheDocument();
    expect(container).toMatchSnapshot();
  });

  it('should render out of stock button correctly', () => {
    render(
      <AddToCartButtonWithProviders
        copy={copy}
        onClick={mockAddToCartButtonOnClick}
        product={product}
        variants={outOfStockVariant}
      />,
    );

    const button = screen.getByRole('button', {
      name: copy.outOfStock.title,
    });

    expect(button).toBeInTheDocument();
  });

  it('should render alternate action button correctly', () => {
    render(
      <AddToCartButtonWithProviders
        copy={copy}
        onClick={mockAddToCartButtonOnClick}
        product={product}
        variants={alternateActionVariant}
      />,
    );

    const button = screen.getByRole('link', {
      name: alternateActionVariant[0].alternateAction.label,
    });

    expect(button).toBeInTheDocument();
  });

  it('should call on click callback successfully', () => {
    render(
      <AddToCartButtonWithProviders
        copy={copy}
        onClick={handleOnClick}
        product={product}
        variants={variants}
      />,
    );

    const button = screen.getByRole('button', {
      name: `${fixture.copy.cartAction} — ${fixture.variants[0].price}`,
    });

    expect(button).toBeInTheDocument();
    expect(handleOnClick).not.toHaveBeenCalled();

    userEvent.click(button);

    expect(handleOnClick).toHaveBeenCalledTimes(1);
  });

  it('should should return null if there are no selected variants', () => {
    render(
      <AddToCartButtonWithProviders
        copy={copy}
        onClick={handleOnClick}
        product={product}
        variants={[]}
      />,
    );

    const button = screen.queryByRole('button');

    expect(button).not.toBeInTheDocument();
  });

  it.todo('should show loading html when clicked');

  it('should have no active timers on unmount (should not throw state update error)', async () => {
    jest.useFakeTimers();

    const { unmount } = render(
      <AddToCartButtonWithProviders
        copy={copy}
        onClick={mockAddToCartButtonOnClick}
        product={product}
        variants={variants}
      />,
    );

    const button = screen.getByRole('button', {
      name: `${fixture.copy.cartAction} — ${fixture.variants[0].price}`,
    });

    userEvent.click(button);

    const successMessage = await screen.findByText(/added to your cart/i);

    expect(successMessage).toBeInTheDocument();

    unmount();

    act(() => {
      jest.runOnlyPendingTimers();
    });

    expect(console.error).not.toHaveBeenCalled(); // eslint-disable-line no-console
  });
});
