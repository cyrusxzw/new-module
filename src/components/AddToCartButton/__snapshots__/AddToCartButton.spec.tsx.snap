// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<AddToCartButton /> should render add to cart button with text to indicate tax inclusion for jp region correctly 1`] = `
<div>
  <button
    class="base alternate blockStyle dark base fullWidth"
    data-test-ref="ADD_TO_CART"
    title="カートに追加する — $26.45 (税込)"
    type="button"
  >
    <span
      class="label"
    >
      <span>
         
      </span>
      <span>
        カートに追加する — $26.45 (税込)
      </span>
    </span>
  </button>
</div>
`;

exports[`<AddToCartButton /> should render base component correctly 1`] = `
<div>
  <button
    class="base alternate blockStyle dark base fullWidth"
    data-test-ref="ADD_TO_CART"
    title="Add to your cart — $26.45"
    type="button"
  >
    <span
      class="label"
    >
      <span>
         
      </span>
      <span>
        Add to your cart — $26.45
      </span>
    </span>
  </button>
</div>
`;
